src/renderer/apiFetch.ts

```diff
@@ -0,0 +1,79 @@
+// ==LICENSE-BEGIN==
+// Copyright 2017 European Digital Reading Lab. All rights reserved.
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import { apiActions } from "readium-desktop/common/redux/actions";
+import { TApiMethod, TApiMethodName } from "readium-desktop/main/api/api.type";
+import { TMethodApi, TModuleApi } from "readium-desktop/main/di";
+import { diRendererGet } from "readium-desktop/renderer/di";
+import { ApiLastSuccess } from "readium-desktop/renderer/redux/states/api";
+import { ReturnPromiseType } from "readium-desktop/typings/pierre";
+import { Unsubscribe } from "redux";
+import * as uuid from "uuid";
+
+export async function apiFetch<T extends TApiMethodName>(path: T, ...requestData: Parameters<TApiMethod[T]>) {
+    return new Promise<ReturnPromiseType<TApiMethod[T]>>((resolve, reject) => {
+        const store = diRendererGet("store");
+        const requestId = uuid.v4();
+        const moduleId = path.split("/")[0] as TModuleApi;
+        const methodId = path.split("/")[1] as TMethodApi;
+        let lastSuccess: ApiLastSuccess | undefined;
+        let storeUnsubscribe: Unsubscribe| undefined;
+        let timeout: number | undefined;
+
+        store.dispatch(
+            apiActions.buildRequestAction(
+                requestId,
+                moduleId,
+                methodId,
+                requestData,
+            ),
+        );
+
+        const promise = new Promise<ReturnPromiseType<TApiMethod[T]>>((resolveSubscribe, rejectSubscribe) => {
+            storeUnsubscribe = store.subscribe(() => {
+                const state = store.getState();
+                const apiLastSuccess = state.api.lastSuccess;
+                const lastSuccessDate = (lastSuccess && lastSuccess.date) || 0;
+
+                if (!apiLastSuccess || apiLastSuccess.date <= lastSuccessDate) {
+                    return;
+                }
+
+                // New api success
+                lastSuccess = apiLastSuccess;
+
+                const meta = apiLastSuccess.action.meta.api;
+                if (moduleId === meta.moduleId && methodId === meta.methodId && state.api.data[requestId]) {
+                    const request = Object.assign({}, state.api.data[requestId]);
+                    store.dispatch(apiActions.clean(requestId));
```

Comment:
In the old `withAPI` HigherOrderComponent (HOC), `apiActions.clean(ID)` was called for each "operationRequest" in the `componentWillUnmount` React lifecycle function. Furthermore, it was possible to dispatch the "clean" action via the component's props, for example in `src/renderer/components/opds/BrowserResult.tsx` immediately before the `apiFetch("opds/browse")`.
Strictly-speaking, the (re)implementation / refactoring in this PR breaks the previous contract, because "clean" is only invoked in the very specific case of `moduleId === meta.moduleId && methodId === meta.methodId && state.api.data[requestId]`. Is that an oversight, or is it intentional? (in which case, I imagine that the "clean" in other scenarios was deemed useless, which justified its removal?)

Reply:
The `clean` redux api data is called before each resolve or reject. In this case the data api in redux state is all the time "clean". When the data is getting back by the promise apiFetch the redux api data state will not useful.

It's false : `because "clean" is only invoked in the very specific case of moduleId === meta.moduleId && methodId === meta.methodId && state.api.data[requestId].`

Sorry if i'm not answering your question, or if i skip anything else

Reply:
I am not sure about exactly what the role of "clean" is (generally-speaking, in the "api" IPC / Redux functionality).

But when I look at the pre-refactoring usage of `cleanData()`:

https://github.com/readium/readium-desktop/blob/6cbd90399ae4c9bcefed0880c8ced776bf57fa5b/src/renderer/components/utils/hoc/api.tsx#L116-L122

...I see that `cleanData()` is automatically called in `componentWillUnmount` of the HOC (which matches the lifecycle of the actual `withApi()` component), and also explicitly called in `browseOpds()`:

https://github.com/readium/readium-desktop/blob/54685000999c2802fa5f3a068f2a1f535eff5fd8/src/renderer/components/opds/BrowserResult.tsx#L105-L118

So, I am trying to understand the post-refactoring lifecycle, to see if this PR introduces breaking changes because of how/when the "clean" functionality is invoked, compared with the old HOC model.

Reply:
> It's false : `because "clean" is only invoked in the very specific case of moduleId === meta.moduleId && methodId === meta.methodId && state.api.data[requestId].`

Well, I am just quoting the exact code from this PR (conditional statement) :)

https://github.com/readium/readium-desktop/blob/6cbd90399ae4c9bcefed0880c8ced776bf57fa5b/src/renderer/apiFetch.ts#L50-L52

Reply:
what is false is `very specific case`in `because "clean" is only invoked in the very specific case`. sorry,

Reply:
before, `cleanData` was called when the component was unmounted or when we called it with `this.props.cleanData` .. now the cleanData is automatically called after each request (apiFetch).
see the redux debugger in web-dev window with `npm run start:dev`

Reply:
we have no side effect. now apiFetch handle the data like an window.fetch method

Reply:
Thank you for the explanation. I trust your judgement, so if you are satisfied with the refactored lifecycle based on your analysis and your tests, then I am happy to merge the proposed change.

Reply:
I'm satisfied of this change..
Thank you Daniel

But before merge I will continue to test all application. To prevent regression bug

Have you make some specific tests ? may be on reader.